Rework ASTDumper

beccadax opened the pull request:
This PR refactors the ASTDumper to make it more structured, less mistake-prone, and more amenable to future changes. For example:

```cpp
  // Before:
  void visitUnresolvedDotExpr(UnresolvedDotExpr *E) {
    printCommon(E, "unresolved_dot_expr")
      << " field '" << E->getName() << "'";
    PrintWithColorRAII(OS, ExprModifierColor)
      << " function_ref=" << getFunctionRefKindStr(E->getFunctionRefKind());
    if (E->getBase()) {
      OS << '\n';
      printRec(E->getBase());
    }
    PrintWithColorRAII(OS, ParenthesisColor) << ')';
  }

  // After:
  void visitUnresolvedDotExpr(UnresolvedDotExpr *E, StringRef label) {
    printCommon(E, "unresolved_dot_expr", label);

    printFieldQuoted(E->getName(), "field");
    printField(E->getFunctionRefKind(), "function_ref", ExprModifierColor);

    if (E->getBase()) {
      printRec(E->getBase());
    }

    printFoot();
  }
```

* Values are printed through calls to base class methods, rather than direct access to the underlying `raw_ostream`.
    * These methods tend to reduce the chances of bugs like missing/extra spaces or newlines, too much/too little indentation, etc.
    * More values are quoted, and unprintable/non-ASCII characters in quoted values are escaped before printing.
* Infrastructure to label child nodes now exists.
    * Some weird breaks from the normal "style", like `PatternBindingDecl`'s original and processed initializers, have been brought into line.
* Some types that previously used ad-hoc dumping functions, like conformances and substitution maps, are now structured similarly to the dumper classes.
* I've fixed the odd dumping bug along the way. For example, distributed actors were only marked `actor`, not `distributed actor`.

This PR doesn't change the overall style of AST dumps; they're still pseudo-S-expressions. But the logic that implements this style is now isolated into a relatively small base class, making it feasible to introduce e.g. JSON dumping in the future.

beccadax committed:
[ASTDumper][NFC] Standardize on getDumpString()

beccadax committed:
[ASTDumper] More getDumpString adoption

beccadax committed:
[ASTDumper] Extract base class for dump visitors

beccadax committed:
[ASTDumper][NFC] Standardize paren printing

ASTDumper now uses new printHead() and printFoot() helpers for this.

beccadax committed:
[ASTDumper][NFC] Make printRec() print newlines

This does an incredible amount of work to clean up the dumper.

beccadax committed:
[ASTDumper] Move Type printRec() into PrintBase

Also changes the signature to be more compatibile with other printRec() overloads.

beccadax committed:
[ASTDumper][NFC] Thread labels through visitors

beccadax committed:
[ASTDumper] Adopt labels in dumps

beccadax committed:
[ASTDumper][NFC] Hoist and add some print helpers

beccadax committed:
[ASTDumper][NFC] Refactor arg/param list handling

beccadax committed:
[ASTDumper][NFC] Add helper for indentation

beccadax committed:
[ASTDumper] Adopt printField

New helper for printing a name-value pair.

beccadax committed:
[ASTDumper][NFC] Add helpers for source locations

beccadax committed:
[ASTDumper] Introduce and adopt printFlag()

Formalizes the printing of unlabeled flags in a node. Also changes the spelling of a few flags, and corrects a bug where distributed actors were always just flagged as `actor`.

beccadax committed:
[ASTDumper] Improve helpers for printing names

beccadax committed:
[ASTDumper] Remove arbitrary I/O from subclasses

Attempts to remove direct access to PrintBase::OS from its subclasses, so that all printing is mediated by the print*() methods.

beccadax committed:
[ASTDumper][NFC] Prevent direct printing

Only PrintBase can access the raw_ostream and indentation directly, so all direct I/O is now isolated to that class. I think this will help us to eventually support multiple output formats (e.g. JSON).

beccadax committed:
[ASTDumper][NFC] Standardize on (value, key) args

printField and several related helpers put the key before the value, while printRec and the visit methods always put the value before the key. Standardize on the latter since we can’t change the visit methods.

beccadax committed:
[ASTDumper][NFC] Omit most blank names

Allows us to more easily audit these later for e.g. JSON support.

beccadax committed:
[ASTDumper] Use PrintBase for conformance dumps

beccadax committed:
[ASTDumper] Update requirement/conformance dumps

Make dumps for Requirement, ProtocolConformance, SubstitionMap, etc. use PrintBase and its helpers.

beccadax committed:
[ASTDumper][NFC] Formatting cleanup, doc comments

beccadax commented:
@swift-ci please smoke test

bnbarham commented on test/expr/capture/generic_params.swift:
```diff
@@ -2,33 +2,33 @@
   // Nested generic functions always capture outer generic parameters, even if
-  // they're not mentioned in the function body
-  // CHECK: func_decl{{.*}}"innerGeneric(u:)" <U> interface type='<T, U> (u: U) -> ()' {{.*}} captures=(<generic> )
+  // they"re not mentioned in the function body
```
This is the only bad replacement I noticed, but I only took a super quick scan (and doesn't really matter)

beccadax replied:
Fixed it. Still perfectly balanced!

beccadax committed:
[ASTDumper] Quote most fields

Except for simple keyword fields and a few other things like capture lists, we now put values in double quotes. We are also correctly escaping these values using functionality built into llvm::raw_ostream.

As part of this change, printField()’s value must now be something that can be handled by a getDumpString() overload. Arbitrary strings should go through printFieldQuoted() or, in a pinch, printFieldRaw().

beccadax commented:
Changes:
* Corrected some commit messages
* Fixed one comment change where an apostrophe was incorrectly changed to a double-quote

beccadax commented:
* More commit message corrections

beccadax commented:
The test jobs from before I started force-pushing are:
* https://ci.swift.org/job/swift-PR-macos-smoke-test/9094/
* https://ci.swift.org/job/swift-PR-Linux-smoke-test/8725/
* https://ci-external.swift.org/job/swift-PR-windows/16511/

beccadax committed:
[ASTDumper] Fix broken test

Updates SILOptimizer/sil_combine_concrete_existential.sil to match updated conformance dumping.

beccadax committed:
[ASTDumper] Escape paths consistently on Windows

Fixes Concurrency/where_clause_main_resolution.swift.

beccadax commented:
@swift-ci please test

ktoso reviewed (approved):
AFAICS nice improvements 👍

AnthonyLatsis commented on lib/AST/ASTDumper.cpp:
```diff
@@ -114,27 +135,62 @@ namespace {
+
+private:
+  virtual void write_impl(const char *Ptr, size_t Size) override {
+    base_os.write_escaped(StringRef(Ptr, Size), /*UseHexEscapes=*/true);
```
FYI, this made the tau in generic parameter names print as escape sequences:
```
(primary_archetype_type address=true name="\xCF\x84_0_0"
  (interface_type=generic_type_param_type depth=0 index=0))
```

beccadax replied:
True. Using `write_escaped` happened to be very convenient, but I wouldn't be averse to a patch that excepted the tau character, or some larger set of known-printable characters (as long as it'll still escape various newlines, quotes, control characters, etc.).
